op-batcher/batcher/driver.go

```diff
@@ -313,6 +319,61 @@ func (l *BatchSubmitter) calculateL2BlockRangeToStore(ctx context.Context) (eth.
 // Submitted batch, but it is not valid
 // Missed L2 block somehow.
 
+func (l *BatchSubmitter) checkLeaderElectionBatcherStatus() (bool, error) {
+	// check if L1Block has changed
+	l1tip, err := l.l1Tip(l.shutdownCtx)
+	if err != nil {
+		l.log.Error("Failed to query L1 tip", "error", err)
+		return false, err
+	}
+	if l.lastL1Tip == l1tip {
+		return l.isLeader, nil
+	}
+
+	sysConf, err := bindings.NewSystemConfigCaller(l.Rollup.L1SystemConfigAddress, l.L1Client)
+	if err == nil {
+		batcherHash, err := sysConf.BatcherHash(&bind.CallOpts{BlockNumber: big.NewInt(int64(l1tip.Number)), Context: l.shutdownCtx})
+		if err == nil {
+			if batcherHash[0] == derive.BatchV1Type {
+				l.Config.BatchInboxVersion = derive.BatchV1Type
+			} else if batcherHash[0] == derive.BatchV2Type {
+				l.Config.BatchInboxVersion = derive.BatchV2Type
+			}
+		}
+	}
+
+	if l.Config.BatchInboxVersion == derive.BatchV1Type {
+		return true, nil
+	}
+
+	lebi, err := bindings.NewLeaderElectionBatchInboxCaller(l.Rollup.BatchInboxContractAddr, l.L1Client)
+	if err != nil {
+		l.log.Error("Failed to set up binding to Batch Inbox Contract", "error", err)
+		return false, err
+	}
+	isCurrentLeader, err := lebi.IsCurrentLeader(&bind.CallOpts{Context: l.shutdownCtx}, l.TxManager.From(), big.NewInt(int64(l1tip.Number+1)))
+	if !isCurrentLeader {
+		l.isLeader = false
+		return false, nil
+	}
+	l.isLeader = true
+
+	_, leaderSpan, err := lebi.NextBlocksAsLeader(&bind.CallOpts{Context: l.shutdownCtx}, l.TxManager.From(), big.NewInt(int64(l1tip.Number+1)))
+	i := 0
+	b := true
+	for i, b = range leaderSpan {
+		if !b {
```

Comment:
Should not it be `if b`, meaning if the bit is true then update the last leader block?

Reply:
Oh... no, but I should make this clearer with a comment.

l1tip is the last completed block.
i=0 means we're on the first block after that.
When we first get b as false, we are at an i that is the position of the first block where we are not leader, relative to l1tip + 1.

So litip + i is the correct block number for last leader block

Reply:
Makes sense!